Nix-style package management

jacobstanley opened the pull request:
Adds support for nix-style package management by creating a sandbox in a global location for each package to be installed, this means they can be shared across sandboxes using symlinks.

The idea comes from https://github.com/hvr/hackage-matrix-builder/ which does something similar.

Not sure it's ready for wider consumption just yet but I want to see if it will build on CI.

jacobstanley committed:
Added dependency on twine

jacobstanley committed:
Nix-style package management

jacobstanley committed:
Add cabal-install version check

jacobstanley committed:
Set MAFIA_HOME for chaos tests

jacobstanley committed:
Check that all the symlinks are intact

jacobstanley committed:
Make sandbox initialisation more obvious

jacobstanley committed:
Remove support for <project>.sandbox files

I don't these would work properly anymore and they don't make sense
anyway now we have a shared global cache.

charleso commented on src/Mafia/Init.hs:
```diff
@@ -2,193 +2,163 @@
+
+------------------------------------------------------------------------
+
+data MafiaState =
```
Worth having a separate module for this?

jacobstanley replied:
Yeah probably, originally it was the only thing in here and was more like `InitState`

jacobstanley replied:
Hmm you know to be fair this is the only thing the init module really does (keep track of the dependency state). Maybe this should just be called something different?

charleso commented on src/Mafia/Install.hs:
```diff
@@ -0,0 +1,214 @@
+
+  env  <- getPackageEnv
+  mapM_ (link packageDB env) globals
+  Hush <- firstEitherT InstallCabalError $ cabal "sandbox" ["hc-pkg", "recache"]
```
Worth adding a comment about what this does?

charleso commented on src/Mafia/Package.hs:
```diff
@@ -48,15 +57,38 @@ packageIdTuple (PackageId n v) =
+  parseLongestMatch Version.parseVersion
+
+parseLongestMatch :: Parse.ReadP a -> Text -> Maybe a
+parseLongestMatch p =
```
Yeah I was surprised this didn't live somewhere already?

jacobstanley replied:
Yeah I'm tempted to rewrite the lot as an attoparsec parser just for consistency

charleso commented on src/Mafia/Install.hs:
```diff
@@ -0,0 +1,214 @@
+  liftIO $ createSymbolicLink (T.unpack pcfg) (T.unpack dest)
+
+install :: PackageEnv -> Package -> EitherT InstallError IO ()
+install env p@(Package (PackageRef pid _ msrc) deps _) = do
```
Some high level doc on this function might be good too

charleso commented on src/Mafia/Cabal/Dependencies.hs:
```diff
@@ -0,0 +1,191 @@
+  List.dropWhile (/= "In order, the following would be installed:") .
+  T.lines
+
+parseDep :: Text -> Either String RevDeps
```
This is probably worth some unit-level tests

charleso commented on src/Mafia/Cabal/Version.hs:
```diff
@@ -0,0 +1,59 @@
+    _                                           -> Nothing
+
+-- TODO not sure if maybe we should have no upper bounds here?
+-- TODO maybe mafia should just install its own cabal?
```
Yeah I guess as discussed that would be the next thing. And then mafia can...

![conquer the world](http://bligoo.com/media/users/0/28040/images/pinky_brain.jpg)

markhibberd replied:
I think installing own cabal is tremendously dangerous/not a good idea. As it stands using cabal and mafia at the same time is standard faire, and risking inconsistent versions could have particularly dire consequences. At least in the naive version of just installing its own and using it regardless of what is on path.

tranma commented on src/Mafia/IO.hs:
```diff
@@ -161,8 +172,29 @@ findExecutable name = liftIO $ do
+mapConcurrentlyE :: Traversable t => (a -> EitherT x IO b) -> t a -> EitherT x IO (t b)
+mapConcurrentlyE io xs = do
+  ys <- liftIO (Async.mapConcurrently (runEitherT . io) xs)
+  hoistEither (sequence ys)
```
nice

jacobstanley committed:
Fix reading of .cabal files with capital letters

Also force constraint to include the version of source dependencies

jacobstanley committed:
Add a few comments on the low level symlinking functionality

jacobstanley committed:
Normalise and sort paths that come out of 'cabal sdist' before hashing

jacobstanley committed:
Added support for .mafiaignore files to control hashing

Files in the `.mafiaignore` will not be included in a source
dependency's hash. This is useful for code generated by a Setup.hs that
may contain timestamps. Such code must be in the sdist in order for
builds to work correctly, but should be ignored for the purposes of
detecting changes to a package.

A 'mafia hash' command is added so that users can easily check whether
their package is being hashed correctly. We could potentially make it a
build failure if calling this twice gives different output.

jacobstanley committed:
Add a .mafiaignore for mafia

jacobstanley committed:
Enable profiling on all libraries installed in the global sandbox

This also adds 'mafia build --enable-profiling'

jacobstanley committed:
Add version number to package cache

jacobstanley committed:
Upgrade libs

jacobstanley committed:
Upgrade QuickCheck

jacobstanley committed:
Add sensible defaults for flags as cabal doesn't have them yet

jacobstanley committed:
Flag is now --profiling

jacobstanley committed:
Added tests for parsing of RevDeps

jacobstanley committed:
Enable documentation by default

jacobstanley committed:
Update test.hs

jacobstanley committed:
Set maximum parallelism to 4x4

jacobstanley committed:
Add environment variables to control parallelism
